File: lib/src/domain/company/entities/tag.entity.dart
import 'package:equatable/equatable.dart';
import 'package:json_annotation/json_annotation.dart';
import 'package:linko/src/domain/company/entities/company.entity.dart';

import '../../user/entities/user.entity.dart';

part 'tag.entity.g.dart';

@JsonSerializable()
class TagEntity extends Equatable {
  int id;
  String? type;
  String name;
  String? desc;
  String? color;
  bool? isDefault;
  bool canDetermine;
  bool isActive;

  List<CompanyEntity>? companies;
  UserEntity? user;

  TagEntity(
      {required this.id,
      required this.name,
      this.type,
      this.desc,
      this.color,
      this.isDefault,
      this.isActive = true,
      this.canDetermine = false,
      this.companies,
      this.user});

  @override
  List<Object?> get props => [
        id,
        name,
        desc,
        type,
        color,
        canDetermine,
        isDefault,
        isActive,
        companies,
        user
      ];

  factory TagEntity.fromJson(Map<String, dynamic> json) =>
      _$TagEntityFromJson(json);

  dynamic get(String propName) {
    switch (propName) {
      default:
        return name;
    }
  }

  Map<String, dynamic> toJson() => _$TagEntityToJson(this);
}

File: src/domain/company/entity/tag.entity.ts
import { User } from "../../user/entity/user.entity";
import { Company } from "./company.entity";
import { TagIdentifier } from "./tag.identifier.entity";

export class Tag {
  id?: number;

  name?: string;

  type?: string;

  desc?: string;

  color?: string;

  searchCount?: number;

  canDetermine?: boolean;

  isActive?: boolean;

  companies?: Company[];

  user?: User;

  tagIdentifiers?: TagIdentifier[];

  constructor(data?: {
    id?: number;
    name?: string;
    type?: string;
    desc?: string;
    color?: string;
    canDetermine?: boolean;
    isActive?: boolean;
    searchCount?: number;
    companies?: Company[];
    user?: User;
    tagIdentifiers?: TagIdentifier[];
  }) {
    this.id = data?.id;
    this.name = data?.name;
    this.type = data?.type;
    this.desc = data?.desc;
    this.searchCount = data?.searchCount;
    this.color = data?.color;
    this.canDetermine = data?.canDetermine;
    this.isActive = data?.isActive;
    this.companies = data?.companies;
    this.user = data?.user;
    this.tagIdentifiers = data?.tagIdentifiers;
  }
}
